pypasta/Repository/Mbox.py

```diff
@@ -374,6 +377,122 @@ def __getitem__(self, message_id):
         return ret
 
 
+class PatchworkProject(MailContainer):
+    NEXT_PAGE_REGEX = re.compile('.*[&?]page=(\d+).*')
+
+    def __init__(self, url, project_id, page_size, d_mbox, f_index, f_mbox_raw):
+        self.url = url
+        self.page_size = page_size
+        self.project_id = project_id
+        self.f_index = f_index
+        self.f_mbox_raw = f_mbox_raw
+        self.d_mbox = d_mbox
+        self.d_mbox_patchwork = os.path.join(self.d_mbox, 'patchwork')
+        self.index = self.load_index(self.f_index)
+        log.info('  ↪ loaded mail index for Patchwork project id %u: found %d mails' %
+                 (project_id, len(self.index)))
+
+
+    def _get_page(self, page):
+        params = dict()
+        params['order'] = 'id'
+        params['project'] = self.project_id
+        params['per_page'] = self.page_size
+        params['page'] = page
+
+        resp = requests.get(urljoin(self.url, 'patches'), params)
+        resp.raise_for_status()
+        json = resp.json()
+
+        page = {int(entry['id']):
+                    (datetime.fromisoformat(entry['date']),
+                     entry['msgid'],
+                     entry['mbox'])
+                for entry in json}
+
+        next_page = None
+        if 'next' in resp.links and 'url' in resp.links['next']:
+            next_url = resp.links['next']['url']
+            match = self.NEXT_PAGE_REGEX.match(next_url)
+            if match:
+                next_page = int(match.group(1))
+
+        return next_page, page
+
+    @staticmethod
+    def _pull_patch(url):
+        resp = requests.get(url)
+        resp.raise_for_status()
+        md5sum = hashlib.md5(resp.content).hexdigest()
+        return md5sum, resp.content
+
+    def update(self):
+        if self.f_mbox_raw:
+            log.info('Processing raw mailbox for Patchwork project id %u' %
+                     self.project_id)
+            process_mailbox_maildir(self.f_mbox_raw, str(self.project_id),
+                           self.d_mbox, mode='patchwork')
+            log.info('  ↪ Initial import successful. You can now remove the initial_archive in the configuration')
+            return
+
+        # Get a set of all present patchwork-ids
+        patchwork_ids = set() \
+            .union(*[{entry[3] for entry in idx_entry}
+                     for idx_entry in self.index.values()])
+        next_page = len(patchwork_ids)//self.page_size + 1
+        pulled = 0
+        error = False
+        while next_page and not error:
+            log.info('Querying page %u' % next_page)
+            try:
+                next_page, page = self._get_page(next_page)
+            except requests.exceptions.RequestException:
+                log.info(' Error fetching page %u. Stopping update' % next_page)
```

Comment:
log.error, as it's an error.

No need to print next_page, we just printed it in line 446. Better print the actual cause of the eception

catch Exception as e:
  log.error(' An error occured during communication with patchwork:')
  log.error(' %s' % str(e))